09-maze_solver/progress/internal/solver/imagefile.go

```diff
@@ -0,0 +1,95 @@
+package solver
+
+import (
+	"fmt"
+	"image"
+	"image/gif"
+	"image/png"
+	"log/slog"
+	"os"
+	"strings"
+)
+
+// openMaze returns a RGBA png image.
+func openMaze(inputPath string) (*image.RGBA, error) {
+	_, err := os.Stat(inputPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to check input file %s: %w", inputPath, err)
+	}
+
+	fd, err := os.Open(inputPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open input image at %s: %w", inputPath, err)
+	}
+	defer fd.Close()
+
+	img, err := png.Decode(fd)
+	if err != nil {
+		return nil, fmt.Errorf("unable to load input image from %s: %w", inputPath, err)
+	}
+
+	// Using RGBAAt() instead of At() saves a lot of time.
+	rgbaImage, ok := img.(*image.RGBA)
+	if !ok {
+		return nil, fmt.Errorf("this isn't a RGBA image")
+	}
+
+	return rgbaImage, nil
+}
+
+// SaveSolution saves the image as a PNG file with the solution path highlighted.
+func (s *Solver) SaveSolution(outputPath string) error {
+	_, err := os.Stat(outputPath)
+	switch {
+	case err == nil:
+		return fmt.Errorf("output file %s already exists", outputPath)
+	case !os.IsNotExist(err):
+		return fmt.Errorf("unable to check output file %s: %w", outputPath, err)
+	}
+
+	fd, err := os.Create(outputPath)
+	if err != nil {
+		return fmt.Errorf("unable to create output image file at %s", outputPath)
+	}
+	defer fd.Close()
```

Comment:
Error of closing is not checked, which may miss a corruption or failure.

Reply:
I've gone with the following, which might be a bit too much. What do you think?.

```go
func (s *Solver) saveSolution(outputPath string) (err error) {
	fd, err := os.Create(outputPath)
	if err != nil {
		return fmt.Errorf("unable to create output image file at %s", outputPath)
	}

	defer func() {
		if closeErr := fd.Close(); closeErr != nil {
			err = errors.Join(err, fmt.Errorf("unable to close file: %w", closeErr))
		}
	}()

	err = png.Encode(fd, s.maze)
	if err != nil {
		return fmt.Errorf("unable to write output image at %s: %w", outputPath, err)
	}

	return nil
}
```

Reply:
Looks perfectly, fine... although, I would also remove the partial file if writing the output fails.

An alternate approach, which doesn't require so much boilerplate, is to write:

```
	var b bytes.Buffer
	err = png.Encode(&b, s.maze)
	if err != nil {
		return fmt.Errorf("unable to write output image at %s: %w", outputPath, err)
	}

	err := os.WriteFile(..., b.Bytes())
...
```

Reply:
PS: however, I do think covering proper error handling for writing a file is important; of course, it doesn't need to happen in this chapter.

Reply:
Keeping Pascal's version to not copy twice the image.